hal/ia32/string.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * HAL basic routines
+ *
+ * Copyright 2012, 2021 Phoenix Systems
+ * Copyright 2001, 2005-2006 Pawel Pisarczyk
+ * Author: Pawel Pisarczyk, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "string.h"
+
+
+void *hal_memcpy(void *dst, const void *src, size_t n)
+{
+	__asm__ volatile(
+		"cld; "
+		"movl %0, %%ecx; "
+		"movl %%ecx, %%edx; "
+		"andl $3, %%edx; "
+		"shrl $2, %%ecx; "
+		"movl %1, %%edi; "
+		"movl %2, %%esi; "
+		"rep; movsl; "
+		"movl %%edx, %%ecx; "
+		"rep; movsb; "
+	:
+	: "g" (n), "g" (dst), "g" (src)
+	: "ecx", "edx", "esi", "edi", "memory", "cc");
+
+	return dst;
+}
+
+
+void *hal_memset(void *dst, int v, size_t n)
+{
+	__asm__ volatile(
+		"cld; "
+		"movl %0, %%ecx; "
+		"movl %%ecx, %%edx; "
```

Comment:
```suggestion
```
In `edx` there is already value `n`.
Now you have something like this:
```
mov %%edx. %%ecx
mov %%ecx, %%edx
```

Reply:
@kemonats thanks for spotting the issue - it's fixed in the next commit.